### PR TITLE
MLIBZ-2122: Allow data validation upfront

### PR DESCRIPTION
**Pull request by heyzooi**
#### Description

Allow data validation strategies upfront before return the lazy results

#### Changes

* Introduces an optional validation in the `FindOperation`

#### Tests

* Coming soon after we agree with the solution

### CODE REVIEW

File: Kinvey/Kinvey/DataStore.swift [hunk 1]
```diff
@@ -103,7 +105,13 @@ open class DataStore<T: Persistable> where T: NSObject {
      - Returns: An instance of `DataStore` which can be a new instance or a cached instance if you are passing a `tag` parameter.
      */
     @available(*, deprecated: 3.0.22, message: "Please use `collection()` instead")
-    open class func getInstance(_ type: StoreType = .cache, deltaSet: Bool? = nil, client: Client = sharedClient, tag: String = defaultTag) -> DataStore {
+    open class func getInstance(
```
Comment (tejasranade) [inline Kinvey/Kinvey/DataStore.swift]: Needs a reference doc for the new `validationStrategy` parameter.

File: Kinvey/Kinvey/DataStore.swift [hunk 2]
```diff
@@ -115,15 +123,30 @@ open class DataStore<T: Persistable> where T: NSObject {
...
      - returns: An instance of `DataStore` which can be a new instance or a cached instance if you are passing a `tag` parameter.
      */
-    open class func collection(_ type: StoreType = .cache, deltaSet: Bool? = nil, autoPagination: Bool = false, client: Client = sharedClient, tag: String = defaultTag) -> DataStore {
+    open class func collection(
+        _ type: StoreType = .cache,
+        deltaSet: Bool? = nil,
+        autoPagination: Bool = false,
+        client: Client = sharedClient,
+        tag: String = defaultTag,
+        validationStrategy: ValidationStrategy? = nil
```
Comment (tejasranade) [inline Kinvey/Kinvey/DataStore.swift]: Same as above. Needs reference doc.

File: Kinvey/Kinvey/ValidationStrategy.swift [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+//  DataValidationStrategy.swift
+//  Kinvey
+//
+//  Created by Victor Hugo Carvalho Barros on 2017-10-12.
+//  Copyright © 2017 Kinvey. All rights reserved.
+//
+
+import Foundation
+
+public enum ValidationStrategy {
```
Comment (tejasranade) [inline Kinvey/Kinvey/ValidationStrategy.swift]: Needs reference doc.

File: Kinvey/Kinvey/ValidationStrategy.swift [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    
+    /// Percentagem between 0.0 and 1.0
+    case randomItems(percentage: Double)
```
Comment (tejasranade) [inline Kinvey/Kinvey/ValidationStrategy.swift]: Can we name this `randomSample`? Usually "sampling" is the term used for what we're doing.

File: Kinvey/Kinvey/ValidationStrategy.swift [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+//
+//  Created by Victor Hugo Carvalho Barros on 2017-10-12.
+//  Copyright © 2017 Kinvey. All rights reserved.
+//
+
+import Foundation
+
+public enum ValidationStrategy {
+    
+    /// Percentagem between 0.0 and 1.0
```
Comment (tejasranade) [inline Kinvey/Kinvey/ValidationStrategy.swift]: Typo.
Also, it'll help to define this better. I suggest saying -
"Percentage between 0.0 and 1.0. This number specifies the fraction of entities in a backend response that are validated. Validating a higher percentage of entities upfront results in a performance penalty."

File: Kinvey/Kinvey/DataStore.swift (hunk 1 above)
Comment (heyzooi) [reply Kinvey/Kinvey/DataStore.swift]: done

File: Kinvey/Kinvey/DataStore.swift (hunk 2 above)
Comment (heyzooi) [reply Kinvey/Kinvey/DataStore.swift]: done

**Pull request merged**